src/format/opamFile.ml

```diff
@@ -385,23 +421,38 @@ module type LineFileArg = sig
   val pp: (string list list, t) Pp.t
 end
 
-module LineFile (X: LineFileArg) = struct
-  module IO = struct
-    include X
+module LineFileOps(X: LineFileArg) = struct
```

Comment:
proposal patch. In the end, i don't know what is better :)
<details><summary>patch</summary>
<p>

```diff

diff --git a/src/format/opamFile.ml b/src/format/opamFile.ml
index f42c88e11..27900b21c 100644
--- a/src/format/opamFile.ml
+++ b/src/format/opamFile.ml
@@ -416,43 +416,29 @@ end
 
 module type LineFileArg = sig
   val internal: string
+  val atomic: bool
   type t
   val empty: t
   val pp: (string list list, t) Pp.t
 end
 
-module LineFileOps(X: LineFileArg) = struct
-  include X
-
-  let format_version = OpamVersion.of_string "0"
+module LineFile(X: LineFileArg) = struct
+  module IO = struct
+    include X
 
-  let to_channel _ oc t = Pp.print (Lines.pp_channel stdin oc -| pp) t
+    let format_version = OpamVersion.of_string "0"
 
-  let to_string _ t = Pp.print (Lines.pp_string -| pp) t
+    let to_channel _ oc t = Pp.print (Lines.pp_channel stdin oc -| pp) t
 
-  let of_channel filename ic =
-    Pp.parse (Lines.pp_channel ic stdout -| pp) ~pos:(pos_file filename) ()
+    let to_string _ t = Pp.print (Lines.pp_string -| pp) t
 
-  let of_string filename str =
-    Pp.parse (Lines.pp_string -| pp)
-      ~pos:{ pos_null with filename = OpamFilename.to_string filename }
-      str
-end
+    let of_channel filename ic =
+      Pp.parse (Lines.pp_channel ic stdout -| pp) ~pos:(pos_file filename) ()
 
-module LineFile (X: LineFileArg) = struct
-  module IO = struct
-    include LineFileOps(X)
-    let atomic = false
-  end
-
-  include IO
-  include MakeIO(IO)
-end
-
-module AtomicLineFile (X: LineFileArg) = struct
-  module IO = struct
-    include LineFileOps(X)
-    let atomic = true
+    let of_string filename str =
+      Pp.parse (Lines.pp_string -| pp)
+        ~pos:{ pos_null with filename = OpamFilename.to_string filename }
+        str
   end
 
   include IO
@@ -467,6 +453,7 @@ end
 module Aliases = LineFile(struct
 
     let internal = "aliases"
+    let atomic = false
 
     type t = string switch_map
 
@@ -485,6 +472,7 @@ module Aliases = LineFile(struct
 module Repo_index (A : OpamStd.ABSTRACT) = LineFile(struct
 
     let internal = "repo-index"
+    let atomic = false
 
     type t = (repository_name * string option) A.Map.t
 
@@ -506,6 +494,7 @@ module Package_index = Repo_index(OpamPackage)
 module PkgList = LineFile (struct
 
     let internal = "package-version-list"
+    let atomic = false
 
     type t = package_set
 
@@ -576,6 +565,7 @@ module Pinned_legacy = struct
   include LineFile(struct
 
       let internal = "pinned"
+      let atomic = false
 
       type t = pin_option OpamPackage.Name.Map.t
 
@@ -592,9 +582,10 @@ end
 
 (** Cached environment updates (<switch>/.opam-switch/environment) *)
 
-module Environment = AtomicLineFile(struct
+module Environment = LineFile(struct
 
     let internal = "environment"
+    let atomic = true
 
     type t = env_update list
 
@@ -627,6 +618,7 @@ module Environment = AtomicLineFile(struct
 module File_attributes = LineFile(struct
 
     let internal = "file_attributes"
+    let atomic = false
 
     type t = file_attribute_set
 
@@ -653,6 +645,7 @@ module File_attributes = LineFile(struct
 module StateTable = struct
 
   let internal = "export"
+  let atomic = false
 
   module M = OpamPackage.Name.Map
 
diff --git a/src/format/opamFile.mli b/src/format/opamFile.mli
index 47759a5eb..b0fca2d87 100644
--- a/src/format/opamFile.mli
+++ b/src/format/opamFile.mli
@@ -1138,6 +1138,7 @@ module SyntaxFile(X: SyntaxFileArg) : IO_FILE with type t := X.t
 
 module type LineFileArg = sig
   val internal: string
+  val atomic: bool
   type t
   val empty: t
   val pp: (string list list, t) OpamPp.t

```
</p>
</details>

Reply:
I started that way, but reverted that to this one just because the diff was less invasive (and it would always be possible to refactor it to allow files to be changed in future when it was needed). So I'm quite happy to do that now!